modules/apps/router/module.go

```diff
@@ -0,0 +1,300 @@
+	// call underlying app's (transfer) callback
+	return am.app.OnChanOpenInit(ctx, order, connectionHops, portID, channelID,
+		chanCap, counterparty, version)
+}
+
+// OnChanOpenTry implements the IBCModule interface
+func (am AppModule) OnChanOpenTry(ctx sdk.Context, order channeltypes.Order, connectionHops []string, portID, channelID string, chanCap *capabilitytypes.Capability, counterparty channeltypes.Counterparty, version, counterpartyVersion string,
+) error {
+	// call underlying app's (transfer) callback
+	return am.app.OnChanOpenTry(ctx, order, connectionHops, portID, channelID,
+		chanCap, counterparty, version, counterpartyVersion)
+}
+
+// OnChanOpenAck implements the IBCModule interface
+func (am AppModule) OnChanOpenAck(ctx sdk.Context, portID, channelID string, counterpartyVersion string) error {
+	return am.app.OnChanOpenAck(ctx, portID, channelID, counterpartyVersion)
+}
+
+// OnChanOpenConfirm implements the IBCModule interface
+func (am AppModule) OnChanOpenConfirm(ctx sdk.Context, portID, channelID string) error {
+	// call underlying app's OnChanOpenConfirm callback.
+	return am.app.OnChanOpenConfirm(ctx, portID, channelID)
+}
+
+// OnChanCloseInit implements the IBCModule interface
+func (am AppModule) OnChanCloseInit(ctx sdk.Context, portID, channelID string) error {
+	// TODO: Unescrow all remaining funds for unprocessed packets
+	return am.app.OnChanCloseInit(ctx, portID, channelID)
+}
+
+// OnChanCloseConfirm implements the IBCModule interface
+func (am AppModule) OnChanCloseConfirm(ctx sdk.Context, portID, channelID string) error {
+	// TODO: Unescrow all remaining funds for unprocessed packets
+	return am.app.OnChanCloseConfirm(ctx, portID, channelID)
+}
+
+// OnRecvPacket implements the IBCModule interface.
+func (am AppModule) OnRecvPacket(ctx sdk.Context, packet channeltypes.Packet, relayer sdk.AccAddress) ibcexported.Acknowledgement {
+	ack := ibcexported.Acknowledgement(channeltypes.NewResultAcknowledgement([]byte{byte(1)}))
+
+	var data transfertypes.FungibleTokenPacketData
+	if err := transfertypes.ModuleCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
+		ack = channeltypes.NewErrorAcknowledgement("cannot unmarshal ICS-20 transfer packet data")
+	}
+
+	// only attempt the application logic if the packet data
+	// was successfully decoded
+	if ack.Success() {
+		receiver, finalDest, port, channel, err := ParseIncomingTransferField(data.Receiver)
+		if finalDest == "" && port == "" && channel == "" && err == nil {
+			ack = am.app.OnRecvPacket(ctx, packet, relayer)
+		} else if err != nil {
+			ack = channeltypes.NewErrorAcknowledgement("cannot unmarshal ICS-20 transfer packet data")
+		} else {
+			if err := am.keeper.OnRecvPacket(ctx, packet, data, receiver, finalDest, port, channel); err != nil {
+				ack = channeltypes.NewErrorAcknowledgement(err.Error())
```

Comment:
Replace this with:

```go
ack = am.app.OnRecvPacket
denom := calculateIntermediateDenom(packet)
newPacket := newTransferPacket(denom, packet.Amount, finalDest)
sendPacket(packet)
```

Reply:
Ok, I'm understanding this now thank you!